### PR TITLE
Reduce compile time of Aggregate

### PR DESCRIPTION
**Pull request by mrks**
This mostly moves around code in the aggregate operator. It is not the oh-so-much needed cleanup/rewrite, but at least reduces the compile time:

Previously (my laptop, debug): 0m35.766s
Now: 0m22.158s

Details:
* Removes unused GroupByContext struct.
* Splits up AggregateContext so that we can use it without the caller being templated on AggregateKey.
* Stores AggregateResult in a vector instead of the AggregateResultMap. This reduces the number of instantiations of that map. Also, it allows for the output step to be independent of AggregateKey.

Performance, just to show that I didn't break anything (not TPC-H, because that is not using the operator heavily enough; measured on nemea):

```
(release)> generate_tpch 10
(release)> SELECT COUNT(*) FROM lineitem GROUP BY l_suppkey LIMIT 10

# Old: EXECUTE: 5 s 847 ms // 5 s 795 ms // 5 s 733 ms
# New: EXECUTE: 5 s 807 ms // 5 s 699 ms // 5 s 590 ms
```

### CODE REVIEW

File: src/lib/operators/aggregate.cpp [hunk 1]
```diff
@@ -23,6 +23,30 @@
...
+  using namespace opossum;
+
+  // Given an AggregateKey key, and a RowId row_id where this AggregateKey was encountered, this first checks if the
+  // AggregateKey was seen before. If not, a new aggregate result is inserted into results and connected to the row id.
+  // This is important so that we can reconstruct the original values later. In any case, a reference to the result is
+  // returned so that result information, such as the aggregate's count or sum, can be modified by the caller.
+  template <typename ResultIds, typename Results, typename AggregateKey>
+  typename Results::reference get_or_add_result(ResultIds& result_ids, Results& results, const AggregateKey& key, const RowID& row_id) {
+    // Get the result id for the current key or add it to the id map
+    auto result_id_inserted = result_ids.try_emplace(key, results.size());
```
Comment (mrks) [inline src/lib/operators/aggregate.cpp]: Interestingly, using `emplace` costs us ~40% of performance.

**Comment by mrks** [comment]
@lanice: Could you run your benchmarks from the monotonic buffer PR again? The vectors are currently not stored in the buffer, but since they use linear storage, I am not sure if that is necessary.

**Comment by lanice** [comment]
Hadn't had the time yet to look into this PR more detailed, but here are the benchmark results, current master (including my change) vs this PR:
```
+-----------+------------------+-------+------------------+-------+--------+---------------------------------+
| Benchmark | prev. iter/s     | runs  | new iter/s       | runs  | change | p-value (significant if <0.001) |
+-----------+------------------+-------+------------------+-------+--------+---------------------------------+
| TPC-H 1   | 9.48307132720947 | 569   | 12.2657508850098 | 736   | +29%   |                          0.0000 |
| TPC-H 2   | 5.56658315658569 | 334   | 5.63251066207886 | 338   | +1%    |                          0.4736 |
| TPC-H 3   | 65.5277938842773 | 3932  | 61.6598663330078 | 3700  | -6%    |                          0.0082 |
| TPC-H 4   | 47.7793617248535 | 2867  | 49.247444152832  | 2955  | +3%    |                          0.1700 |
| TPC-H 5   | 62.3465003967285 | 3741  | 63.2133522033691 | 3793  | +1%    |                          0.1997 |
| TPC-H 6   | 440.869354248047 | 10000 | 439.806976318359 | 10001 | -0%    |                          0.0678 |
| TPC-H 7   | 39.580394744873  | 2375  | 40.515022277832  | 2432  | +2%    |                          0.0002 |
| TPC-H 8   | 17.6871318817139 | 1062  | 17.6453151702881 | 1059  | -0%    |                          0.3006 |
| TPC-H 9   | 35.8127250671387 | 2149  | 37.894115447998  | 2274  | +6%    |                          0.0003 |
| TPC-H 10  | 77.2794876098633 | 4637  | 74.8647079467773 | 4492  | -3%    |                          0.0000 |
| TPC-H 11  | 904.392395019531 | 10010 | 952.10791015625  | 10008 | +5%    |                          0.0000 |
| TPC-H 12  | 268.045806884766 | 10000 | 268.1484375      | 10001 | +0%    |                          0.9404 |
| TPC-H 13  | 24.899923324585  | 1494  | 14.749361038208  | 885   | -41%   |                          0.0000 |
| TPC-H 14  | 502.785888671875 | 10001 | 506.671508789062 | 10003 | +1%    |                          0.0188 |
| TPC-H 16  | 51.0632400512695 | 3064  | 52.9981956481934 | 3180  | +4%    |                          0.0000 |
| TPC-H 18  | 32.2972068786621 | 1938  | 22.1995620727539 | 1332  | -31%   |                          0.0000 |
| TPC-H 19  | 319.277770996094 | 10000 | 334.804626464844 | 10005 | +5%    |                          0.0000 |
| TPC-H 22  | 416.774322509766 | 10000 | 412.055450439453 | 10004 | -1%    |                          0.0820 |
| average   |                  |       |                  |       | -1%    |                                 |
+-----------+------------------+-------+------------------+-------+--------+---------------------------------+
```

**Comment by mrks** [comment]
Also, gcc or clang?

**Comment by lanice** [comment]
> Is this correct:

Yes, with scale factor 1 (or is this the default now?). Also, I left out queries 17,20,21, because they are pushing the execution time so much with 100 clients, and I wanted to have the results before the next day ;)

> Also, gcc or clang?

I believe clang

**Comment by mrks** [comment]
Now:

```
| Benchmark | prev. iter/s  | runs  | new iter/s    | runs  | change | p-value (significant if <0.001) |
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
| TPC-H 1   | 15.9987716675 | 960   | 12.5492458344 | 753   | -22%   |                          0.0000 |
| TPC-H 2   | 5.89935398102 | 354   | 6.1826672554  | 371   | +5%    |                          0.4477 |
| TPC-H 3   | 68.8749084473 | 4133  | 68.490196228  | 4110  | -1%    |                          0.8288 |
| TPC-H 4   | 45.7643432617 | 2746  | 51.71068573   | 3103  | +13%   |                          0.0000 |
| TPC-H 5   | 59.2104110718 | 3553  | 65.9138717651 | 3955  | +11%   |                          0.0000 |
| TPC-H 6   | 453.210083008 | 10000 | 463.758575439 | 10001 | +2%    |                          0.0000 |
| TPC-H 7   | 31.8477039337 | 1911  | 41.2496604919 | 2475  | +30%   |                          0.0000 |
| TPC-H 8   | 18.911151886  | 1135  | 19.2568340302 | 1156  | +2%    |                          0.2828 |
| TPC-H 9   | 35.4976882935 | 2130  | 39.2272758484 | 2354  | +11%   |                          0.0000 |
| TPC-H 10  | 82.7401046753 | 4965  | 82.8384246826 | 4971  | +0%    |                          0.1309 |
| TPC-H 11  | 990.130004883 | 10006 | 954.718688965 | 10007 | -4%    |                          0.0000 |
| TPC-H 12  | 281.468048096 | 10001 | 297.35333252  | 10000 | +6%    |                          0.0000 |
| TPC-H 13  | 18.6318798065 | 1118  | 14.6310377121 | 878   | -21%   |                          0.0000 |
| TPC-H 14  | 545.170959473 | 10003 | 565.148132324 | 10005 | +4%    |                          0.0000 |
| TPC-H 16  | 34.6303443909 | 2078  | 59.2283859253 | 3554  | +71%   |                          0.0000 |
| TPC-H 18  | 30.863368988  | 1852  | 34.4618606567 | 2068  | +12%   |                          0.0000 |
| TPC-H 19  | 343.730651855 | 10000 | 348.419769287 | 10000 | +1%    |                          0.0305 |
| TPC-H 22  | 416.337127686 | 10008 | 422.568634033 | 10007 | +1%    |                          0.0543 |
| average   |               |       |               |       | +7%    |                                 |
+-----------+---------------+-------+---------------+-------+--------+---------------------------------+
```

**Comment by Bouncner** [comment]
Since 1 and 13 are "slower" than 16 (which gained a lot) already: any ideas why we lose performance here?

**Comment by mrks** [comment]
Instead of storing the result struct in the map, we have an additional indirection into a vector. That seems to be good for many queries, but not so good for others. I would agree with @lanice and accept those changes for now.

**Comment by mrks** [comment]
Might also be random variations, because TPC-H 1 was not as bad in the first benchmark.

**Pull request merged**